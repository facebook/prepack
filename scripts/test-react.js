/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

/* @flow */

let fs = require("fs");
let path = require("path");
let { prepackSources } = require("../lib/prepack-node.js");
let babel = require("babel-core");
let React = require("react");
let ReactTestRenderer = require("react-test-renderer");
let { mergeAdacentJSONTextNodes } = require("../lib/utils/json.js");
/* eslint-disable no-undef */
let { expect, describe, it } = global;

let reactTestRoot = path.join(__dirname, "../test/react/");
let prepackOptions = {
  compatibility: "react-mocks",
  errorHandler: diag => "Fail",
  internalDebug: true,
  serialize: true,
  uniqueSuffix: "",
  maxStackDepth: 100,
  reactEnabled: true,
  inlineExpressions: true,
  omitInvariants: true,
};

function compileSourceWithPrepack(source) {
  let code = `(function(){${source}})()`;
  let serialized = prepackSources([{ filePath: "", fileContents: code, sourceMapContents: "" }], prepackOptions);
  if (serialized == null || serialized.reactStatistics == null) {
    throw new Error("React test runner failed during serialization");
  }
  return {
    compiledSource: serialized.code,
    statistics: serialized.reactStatistics,
  };
}

function runSource(source) {
  let codeAfterBabel = babel.transform(source, {
    presets: ["babel-preset-react"],
    plugins: ["transform-object-rest-spread"],
  }).code;
  /* eslint-disable no-new-func */
  let fn = new Function("require", "module", codeAfterBabel);
  let moduleShim = { exports: null };
  let requireShim = name => {
    switch (name) {
      case "react":
        return React;
      default:
        throw new Error(`Unrecognized import: "${name}".`);
    }
  };
  let global = {
    require: requireShim,
    module: moduleShim,
    Object,
    String,
  };
  try {
    // $FlowFixMe flow doesn't new Function
    fn.call(global, requireShim, moduleShim);
  } catch (e) {
    console.log(codeAfterBabel);
    throw e;
  }
  return moduleShim.exports;
}

async function runTest(directory, name) {
  let source = fs.readFileSync(path.join(reactTestRoot, directory, name)).toString();
  let { compiledSource } = compileSourceWithPrepack(source);

  let A = runSource(source);
  expect(typeof A).toBe("function");
  let B = runSource(compiledSource);
  expect(typeof B).toBe("function");

  let rendererA = ReactTestRenderer.create(null);
  let rendererB = ReactTestRenderer.create(null);

  if (A == null || B == null) {
    throw new Error("React test runner issue");
  }
  // // Use the original version of the test in case transforming messes it up.
  let { getTrials } = A;
  // // Run tests that assert the rendered output matches.
  let resultA = getTrials(rendererA, A);
  let resultB = getTrials(rendererB, B);

  // // the test has returned many values for us to check
  for (let i = 0; i < resultA.length; i++) {
    let [nameA, valueA] = resultA[i];
    let [nameB, valueB] = resultB[i];
    expect(mergeAdacentJSONTextNodes(valueB)).toEqual(mergeAdacentJSONTextNodes(valueA));
    expect(nameB).toEqual(nameA);
  }
}

// Jest tests
let originalConsoleError = console.error;

describe("Test React", () => {
  describe("Functional component folding", () => {
    let directory = "functional-components";

    it("Simple", async () => {
      await runTest(directory, "simple.js");
    });

    it("Simple children", async () => {
      await runTest(directory, "simple-children.js");
    });

    it("Simple refs", async () => {
      await runTest(directory, "simple-refs.js");
    });

    it("Conditional", async () => {
      await runTest(directory, "conditional.js");
    });

    it("Key nesting", async () => {
      await runTest(directory, "key-nesting.js");
    });

    it("Key change", async () => {
      await runTest(directory, "key-change.js");
    });

    it("Component type change", async () => {
      await runTest(directory, "type-change.js");
    });

    it("Component type same", async () => {
      await runTest(directory, "type-same.js");
    });

    it("Dynamic props", async () => {
      await runTest(directory, "dynamic-props.js");
    });

    it("Dynamic context", async () => {
      await runTest(directory, "dynamic-context.js");
    });

    it("React.cloneElement", async () => {
      await runTest(directory, "clone-element.js");
    });

    it("Return text", async () => {
      await runTest(directory, "return-text.js");
    });

    it("Return undefined", async () => {
      // this test will cause a React console.error to show
      // we monkey patch it to stop it polluting the test output
      // with a false-negative error
      global.console.error = () => {};
      try {
        await runTest(directory, "return-undefined.js");
      } finally {
        global.console.error = originalConsoleError;
      }
    });
  });
});
