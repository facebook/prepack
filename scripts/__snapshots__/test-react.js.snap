// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test React (JSX) Class component folding Classes with state 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Class component folding Inheritance chaining 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (JSX) Class component folding Simple classes #2 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Class component folding Simple classes #3 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (JSX) Class component folding Simple classes 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Factory class component folding Simple factory classes 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Factory class component folding Simple factory classes 2 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Additional functions closure scope capturing 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Circular reference 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Class component as root 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Class component as root with instance variables #2 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Class component as root with instance variables 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Class component as root with multiple render methods 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Class component as root with props 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Class component as root with refs 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Class component as root with state 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Component type change 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 2,
}
`;

exports[`Test React (JSX) Functional component folding Component type same 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Conditional 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Dynamic context 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Dynamic props 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Key change 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (JSX) Functional component folding Key change with fragments 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (JSX) Functional component folding Key nesting 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 2,
}
`;

exports[`Test React (JSX) Functional component folding Key nesting 2 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Key nesting 3 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Key not changing with fragments 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (JSX) Functional component folding React.cloneElement 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Return text 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Return undefined 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Simple 1`] = `
ReactStatistics {
  "inlinedComponents": 3,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Simple 2 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Simple 3 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Simple 4 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (JSX) Functional component folding Simple 5 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 3,
}
`;

exports[`Test React (JSX) Functional component folding Simple 6 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Simple children 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Simple fragments 1`] = `
ReactStatistics {
  "inlinedComponents": 3,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) Functional component folding Simple refs 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) fb-www mocks fb-www 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) fb-www mocks fb-www 2 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) fb-www mocks fb-www 3 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 0,
}
`;

exports[`Test React (JSX) fb-www mocks fb-www 4 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 0,
}
`;

exports[`Test React (JSX) fb-www mocks fb-www 5 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 0,
}
`;

exports[`Test React (JSX) fb-www mocks fb-www 6 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 0,
}
`;

exports[`Test React (JSX) fb-www mocks fb-www 7 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (JSX) fb-www mocks fb-www 8 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (create-element) Class component folding Classes with state 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Class component folding Inheritance chaining 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (create-element) Class component folding Simple classes #2 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Class component folding Simple classes #3 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (create-element) Class component folding Simple classes 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Factory class component folding Simple factory classes 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Factory class component folding Simple factory classes 2 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Additional functions closure scope capturing 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Circular reference 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Class component as root 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Class component as root with instance variables #2 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Class component as root with instance variables 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Class component as root with multiple render methods 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Class component as root with props 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Class component as root with refs 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Class component as root with state 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Component type change 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 2,
}
`;

exports[`Test React (create-element) Functional component folding Component type same 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Conditional 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Dynamic context 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Dynamic props 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Key change 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (create-element) Functional component folding Key change with fragments 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (create-element) Functional component folding Key nesting 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 2,
}
`;

exports[`Test React (create-element) Functional component folding Key nesting 2 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Key nesting 3 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Key not changing with fragments 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (create-element) Functional component folding React.cloneElement 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Return text 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Return undefined 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Simple 1`] = `
ReactStatistics {
  "inlinedComponents": 3,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Simple 2 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Simple 3 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Simple 4 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;

exports[`Test React (create-element) Functional component folding Simple 5 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 3,
}
`;

exports[`Test React (create-element) Functional component folding Simple 6 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Simple children 1`] = `
ReactStatistics {
  "inlinedComponents": 2,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Simple fragments 1`] = `
ReactStatistics {
  "inlinedComponents": 3,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) Functional component folding Simple refs 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) fb-www mocks fb-www 1`] = `
ReactStatistics {
  "inlinedComponents": 1,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) fb-www mocks fb-www 2 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) fb-www mocks fb-www 3 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 0,
}
`;

exports[`Test React (create-element) fb-www mocks fb-www 4 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 0,
}
`;

exports[`Test React (create-element) fb-www mocks fb-www 5 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 0,
}
`;

exports[`Test React (create-element) fb-www mocks fb-www 6 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 0,
}
`;

exports[`Test React (create-element) fb-www mocks fb-www 7 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 1,
}
`;

exports[`Test React (create-element) fb-www mocks fb-www 8 1`] = `
ReactStatistics {
  "inlinedComponents": 0,
  "optimizedTrees": 2,
}
`;
